### PR TITLE
#808 Add Cors feature

### PR DESCRIPTION
**Pull request by rserj**
Feature allows to make Cross Origin Requests for Open Id, please see #808 ticket for more information

### CODE REVIEW

File: src/OrchardCore.Modules/Orchard.OpenId/Controllers/AccessController.cs [hunk 1]
```diff
@@ -24,6 +26,7 @@
 namespace Orchard.OpenId.Controllers
 {
     [Authorize]
+    [EnableCors(Constants.OpenIdConnectPolicy)]
```
Comment (anoordende) [inline src/OrchardCore.Modules/Orchard.OpenId/Controllers/AccessController.cs]: Is there a way of restricting this to only those actions that may require CORS?

File: src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdCorsConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+        public void Configure(CorsOptions options)
+        {
+            var openIdSettings = _openIdService.GetOpenIdSettingsAsync().GetAwaiter().GetResult();
+            if (openIdSettings.Audiences != null && openIdSettings.Audiences.Any())
+                options.AddPolicy(Constants.OpenIdConnectPolicy, builder => builder
+                    .WithOrigins(openIdSettings.Audiences.ToArray())
+                    .AllowAnyHeader()
+                    .AllowAnyMethod()
+                    .AllowCredentials()
```
Comment (anoordende) [inline src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdCorsConfiguration.cs]: If this is required for any of the flows, we should consider configuring the policy based on the enabled flows.

File: src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdCorsConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+            _openIdService = openIdService;
+        }
+
+        public void Configure(CorsOptions options)
+        {
+            var openIdSettings = _openIdService.GetOpenIdSettingsAsync().GetAwaiter().GetResult();
+            if (openIdSettings.Audiences != null && openIdSettings.Audiences.Any())
+                options.AddPolicy(Constants.OpenIdConnectPolicy, builder => builder
+                    .WithOrigins(openIdSettings.Audiences.ToArray())
+                    .AllowAnyHeader()
```
Comment (anoordende) [inline src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdCorsConfiguration.cs]: Most restrictive policy should be applied, ie. only the verbs and headers that are required, something along the lines of GET, POST, OPTIONS and accept, content_type and authorization.

File: src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdCorsConfiguration.cs (hunk 2 above)
Comment (rserj) [inline src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdCorsConfiguration.cs]: Hmm, I didn't think about flows. 
If Admin set some external Audience, He assuming that External host can connect to API. However, I missed one case, where Audience will be the same as current host

>If you are providing tokens for accessing resources on this site Audiences should be equal to Authority url

Probably we should just check if Audience equals to current Host then ignore it?

File: src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdCorsConfiguration.cs (hunk 3 above)
Comment (rserj) [inline src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdCorsConfiguration.cs]: Agree, Will do

File: src/OrchardCore.Modules/Orchard.OpenId/Controllers/AccessController.cs (hunk 1 above)
Comment (rserj) [inline src/OrchardCore.Modules/Orchard.OpenId/Controllers/AccessController.cs]: Yeh, I will apply it per Action which requires CORS, thank you

File: src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdCorsConfiguration.cs (hunk 2 above)
Comment (anoordende) [inline src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdCorsConfiguration.cs]: Flows based on redirects don't require Cors for the `AccessController`, only for the `UserInfo` controller. We are also assuming that every configured App requires UserInfo. So, we'd have to enable/disable Cors at the App level (or for each audience)  and have 2 policies: one for the AccessController and another for the UserController. Also, not all flows require `AllowCredentials`.

**Comment by PinpointTownes** [comment]
Cool stuff happening here! :+1: 

Two remarks, tho':

  - You shouldn't use `OpenIdSettings.Audiences` for the trusted origins. `Audiences` is only supposed to contain the list of "resources" a client application is allowed to query, not the URLs of the clients that act as JS HTTP clients (it should probably renamed to `Resources`). Consider using a separate list. Alternatively, we could consider adding the list directly under each application, and resolve the trusted origins dynamically, using `OpenIdApplicationStore`.

  - Doing CORS at the controller/action level (with `[EnableCors]`) is likely not the best option for at least 3 reasons:

    - API endpoints like the discovery endpoints (configuration + JWKS) are directly managed by OpenIddict and don't have dedicated controllers/actions. Yet, JS apps should be able to use them.

    - Even if you have an action for the token endpoint, requests that are directly rejected by OpenIddict (e.g because they lack a required parameter, or because client authentication failed) won't get the CORS response headers and the JS apps that receive errored responses won't be able to read them.

    - OpenIddict (well, the OIDC server behind OpenIddict to be exact) is quite zealous and will automatically reject HTTP requests that don't use GET or POST (depending on the endpoint). Concretely, this means that preflight requests will be rejected by OpenIddict before they have a chance to be handled by MVC's CORS filter.

For these reasons, you should instead consider implementing CORS at the middleware level.

/cc @sebastienros

File: src/OrchardCore.Modules/Orchard.OpenId/Controllers/AccessController.cs (hunk 1 above)
Comment (PinpointTownes) [inline src/OrchardCore.Modules/Orchard.OpenId/Controllers/AccessController.cs]: I agree with @anoordende. We need to make sure CORS is only applied to non-interactive endpoints, which excludes both the authorization and logout endpoints.

**Comment by rserj** [comment]
@PinpointTownes Thanks for you comments
>API endpoints like the discovery endpoints (configuration + JWKS) are directly managed by OpenIddict and don't have dedicated controllers/actions. Yet, JS apps should be able to use them.

Yes, I agree I also mentioned it in #808 ticket. It will be better adding "Origins" per application. 
But I have small concern, I have to refresh CorsOptions somehow, each time after Creating/Updating Application Unless whole Tenant need tobe restarAdmin in order to changes take effect.

>Doing CORS at the controller/action level (with [EnableCors]) is likely not the best option for at least 3 reasons:

I'm not an expert in this area but, i'll try to comment
>API endpoints like the discovery endpoints (configuration + JWKS) are directly managed by OpenIddict and don't have dedicated controllers/actions. Yet, JS apps should be able to use them.

I do not know how openiddict handle CORS for "discovery end points", I tried to make cross-origin request to configuration end-point ".well-known/openid-configuration", it worked fine.
Probably it may fail in production?

>Even if you have an action for the token endpoint, requests that are directly rejected by OpenIddict (e.g because they lack a required parameter, or because client authentication failed) won't get the CORS response headers and the JS apps that receive errored responses won't be able to read them.

Yes we have separate Action for Token end point. I agree that it will be hard to determine real reason of failed request, but Preflight request should solve it right?

>OpenIddict (well, the OIDC server behind OpenIddict to be exact) is quite zealous and will automatically reject HTTP requests that don't use GET or POST (depending on the endpoint). Concretely, this means that preflight requests will be rejected by OpenIddict before they have a chance to be handled by MVC's CORS filter.
Do you mean discovery endpoints? We use own separate Controllers/Actions for UserInfo/Authorization/Token end points, I guess Preflight requests should work since those are our end points ?

I see couple major disadvantages in implementing CORS as middleware.
-You can not seletivly choose 'Cors Policy' for individual Controller/Action.
-We should write our own middleware for such case, cause standart implementation allows to use only one Policy per middleware

**Comment by anoordende** [comment]
Thanks @PinpointTownes, great explanation.

That does raise a very fundamental question, unless I'm mistaken, this means throughout Orchard we cannot use MVC Cors (they are mutually exclusive). By default, we can only apply a single policy using the middleware `UseCors`, so we'd need to implement `ICorsPolicyProvider` to select the policy based on the `HttpContext` route dictionary.

**Thinking out loud:**
Implement `ICorsPolicyProvider` in the new Cors module, for example an `AppSettingsCorsPolicyProvider` that loads the policies from appsettings:

~~~~
    "Cors": {
        "Policies": [
            {
                "Name": "Orchard.OpenId.UserInfo",
                "Origins": [ "https://my.externaldomain.com" ],
                "Methods": [ "GET", "POST", "OPTIONS" ],
                "Headers": [ "accept", "content-type", "authorization" ],
                "PreflightMaxAge": 86400
            },
            {
                "Name": "Orchard.SomeOtherModule.SomeController.Index",
                "Origins": [ "https://my.externaldomain.com" ],
                "Methods": [ "GET", "POST", "PATCH", "OPTIONS" ],
                "Headers": [ "accept", "content-type", "authorization" ],
                "AllowCredentials": true
                "PreflightMaxAge": 43200
            }
        ]
    }
~~~~

And something along the lines of:

~~~~
        public static IServiceCollection AddCors(this IServiceCollection services, IConfiguration configuration) {
            var cors = new CorsOptions();
            configuration.GetSection("Cors").Bind(cors);
            services.AddCors(options => {
                foreach (var policy in cors.Policies) {
                    options.AddPolicy(policy.Name, config => {
                        config
                            .WithOrigins(policy.Origins)
                            .WithMethods(policy.Methods)
                            .WithHeaders(policy.Headers)
                            .SetPreflightMaxAge(TimeSpan.FromSeconds(policy.PreflightMaxAge));
                    });
                }
                options.DefaultPolicyName = cors.DefaultPolicy;
            });
            return services;
        }
~~~~

The naming of the policies can follow a pattern that identifies the requests to apply each policy on, increasingly more specific (or conversely, more generic), based on Area, Controller, Action. For example, the policy provider would look for a policy for `/Orchard.OpenId/UserInfo/Me` like this:

- Orchard.OpenId.UserInfo.Me
- Orchard.OpenId.UserInfo
- Orchard.OpenId
- Orchard
- Default
- (Deny)

Lastly, having gone full circle, I'm really not so sure that the OpenId module should generate these policies dynamically. I'd much prefer a single point of truth and very granular control instead of assumptions by modules.

**Comment by rserj** [comment]
If we decided stick with Middleware approach, probably we should use MVC RouteTemplate (eg {controller}/{Action}/{arg1}/{arg2}) instead of "Name"? 
In such case Do we need to fall-back request to the parent policy if current policy's route didn't mach?
(like, ActionPolicy->ControllerPolicy->Deny) 

Basicaly it means we are trying implement the same filtration approach which we have in MVC but in middleware level. 

**I see two big conserns**
1. Every time Admin, creates, update OpenId Application he should also go to the configuration file and manualy change origins.
2. Unlike MVC Filters ([EnableCors]), Middleware will be executing in each request, we may have perfomance problems

I'm not an export in OpenIddict but I tried "Password flow" with MVC Origin Filtration ([EnableCors]) with Cross Origin requests, I could retrieve Token and Discovery End-points worked as well.
Probably we should stick With MVC Filtration approach? All we have todo add Origins configuration per OpenId Application.

**Comment by anoordende** [comment]
I just made a quick PoC in one of our internal projects using the `ICorsPolicyProvider`, completely removing attributes and dynamically selecting the policy, albeit using policy names composed of request path + method instead of Area/Controller/Action. It does provide the granularity and control I'm seeking.

This is an alternative approach to the current PR, but if anyone's interested, I can put this in a separate PR over the next few days so we can compare the two, pick & mix and arrive at a solution to move forward with. Simply 👍  or 👎 me here :)

**Comment by anoordende** [comment]
PS: Sorry @rserj our comments got cross-wired. 

- Yes, it needs to fall back all the way to denying the Cors request.
- Once the settings-driven approach is implemented, which I feel will benefit all of Orchard, we can add the "dynamic" per-app origins from OpenId (and any other module that needs it), though probably the AppSettings would always override. To me this is just an administrative nice-to-have, but if it's important to you, then it may be important to others.
- I doubt there's much of a noticeable difference in performance between declarative and programmatic, after all, it uses the same internals. But this we can measure :)

**Comment by rserj** [comment]
I added AllowedOrigins Property to OpenId Application and made some changes according to your notes.
Please take a look, thanks

**Comment by rserj** [comment]
I have one concern here, In case if I made changes to OpenId Application's AllowedOrigins, in order to changes take effect, I have to restart Tenant, I guess I have to Warn about it as we do in global OpenId settings by showing Warning message, what do you think?

File: src/OrchardCore.Modules/Orchard.OpenId/ModelBinders/UrlsAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    {
+
+        public Task BindModelAsync(ModelBindingContext bindingContext)
+        {
+            var valueProviderResult = bindingContext.ValueProvider.GetValue(bindingContext.FieldName);
+            if (valueProviderResult.Length > 0)
+            {
+                var valueAsString = valueProviderResult.FirstValue;
+
+                if (string.IsNullOrEmpty(valueAsString))
```
Comment (Ermesx) [inline src/OrchardCore.Modules/Orchard.OpenId/ModelBinders/UrlsAttribute.cs]: This `if` could be removed.

File: src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdCorsConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        public void Configure(CorsOptions options)
+        {
+            var openIdApplications = _openIdApplicationStore.GetAllApps().GetAwaiter().GetResult();
+            if (openIdApplications == null || !openIdApplications.Any()) return;
+            var appOrigins = openIdApplications
+                .Where(app => (app.AllowPasswordFlow || app.AllowClientCredentialsFlow || app.AllowRefreshTokenFlow) 
+                              && app.AllowedOrigins != null && app.AllowedOrigins.Any())
+                .SelectMany(app => app.AllowedOrigins)
+                .ToArray();
+            if (!appOrigins.Any()) return;
```
Comment (Jetski5822) [inline src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdCorsConfiguration.cs]:
```c#
if (appOrigins.Count > 0) 
{
return;
}
```

File: src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdCorsConfiguration.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+        public OpenIdCorsConfiguration(OpenIdApplicationStore openIdService)
+        {
+            _openIdApplicationStore = openIdService;
+        }
+
+        public void Configure(CorsOptions options)
+        {
+            var openIdApplications = _openIdApplicationStore.GetAllApps().GetAwaiter().GetResult();
+            if (openIdApplications == null || !openIdApplications.Any()) return;
```
Comment (Jetski5822) [inline src/OrchardCore.Modules/Orchard.OpenId/Services/OpenIdCorsConfiguration.cs]: .Count == 0

File: src/OrchardCore/OrchardCore.Cms/OrchardCore.Cms.csproj [hunk 7]
```diff
@@ -22,6 +22,7 @@
   -->
 
   <ItemGroup>
+    <ProjectReference Include="..\..\OrchardCore.Modules\Orchard.Cors\Orchard.Cors.csproj" />
```
Comment (Jetski5822) [inline src/OrchardCore/OrchardCore.Cms/OrchardCore.Cms.csproj]: Can you add PrivateAssets="none" here please

File: src/OrchardCore.Modules/Orchard.OpenId/ModelBinders/UrlsAttribute.cs (hunk 4 above)
Comment (rserj) [inline src/OrchardCore.Modules/Orchard.OpenId/ModelBinders/UrlsAttribute.cs]: removed, thanks

**Comment by rserj** [comment]
Nicholas and Wojciech I made commit regarding to your reviews, thanks

**Comment by anoordende** [comment]
Sorry, holidays. Will get back to this thread shortly.

**Comment by rserj** [comment]
I'll close this PR, please see discussion in 
https://github.com/OrchardCMS/OrchardCore/issues/854

**Pull request closed without merging**